### PR TITLE
[CELEBORN-1286] Introduce configuration.md to document dynamic config and config service

### PR DESCRIPTION
**Pull request by SteNicholas**
### What changes were proposed in this pull request?

Introduce `configuration.md` to document dynamic config and config service.

### Why are the changes needed?

`DynamicConfig` and `ConfigService` have already been supported in #2100, which should be documented to introduce the feature.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

No.

### CODE REVIEW

**Comment by SteNicholas** [comment]
cc @RexXiong, @AngersZhuuuu.

File: docs/developers/configuration.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+## Config Service
+Config service provides the configuration management service with local cache for the static and dynamic configuration. 
+Meanwhile, `ConfigService` is a pluggable service interface of which implementation is based on various store backend. 
+The store backend of `ConfigService` is configured with `celeborn.dynamicConfig.store.backend`, which supports filesystem 
+and database store backend at present. If the store backend is not provided, it means that config service is disabled. 
+
+### FileSystem Config Service
+Filesystem config service supports the configuration specified in the dynamic configuration file. The dynamic configuration 
+file is configured with `celeborn.quota.configuration.path`, of which default path is `$CELEBORN_HOME/conf/dynamicConfig.yaml`. 
+The template of the dynamic configuration is as follows:
```
Comment (AngersZhuuuu) [inline docs/developers/configuration.md:50]: Add example section?

File: docs/developers/configuration.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+       config:
+         celeborn.client.push.buffer.initial.size: 1k
+         celeborn.client.push.buffer.initial.size.user.only: 512k
+```
+
+### Database Config Service
+Database config service refreshes the changed dynamic configuration stored in database via JDBC. The configuration of 
+database store backend is specified with `celeborn.dynamicConfig.store.db.*`. The database of store backend needs to create 
+the tables of the dynamic configuration for different config levels.  For example, the template of configuration tables for 
+MySQL is as follows:
```
Comment (AngersZhuuuu) [inline docs/developers/configuration.md:94]: Add a example section?
Also we may need a section about db related setting

File: docs/developers/configuration.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+---
+
+# Configuration
+Celeborn configuration is separated into static and dynamic configuration. Static configuration `CelebornConf` is specified in 
+default configuration file `$CELEBORN_HOME/conf/celeborn-defaults.conf`. Meanwhile, the dynamic configuration overrides
```
Comment (AngersZhuuuu) [inline docs/developers/configuration.md:21]: ## Static Configuration
Static configuration `CelebornConf` is specified in default configuration file `$CELEBORN_HOME/conf/celeborn-defaults.conf`. 

## Dynamic Configuration
The dynamic configuration overrides the static configuration at runtime of `Master` and `Worker`.

File: docs/developers/configuration.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+When applying dynamic configuration, the priority order of config levels is as follows: 
+
+- `SYSTEM` level configuration overrides the static configuration, which in turn overrides the configuration of `CelebornConf`. 
+When the system level configuration is missing, fallback to the static configuration of `CelebornConf`.
+- `TENANT` level configuration overrides `SYSTEM` level, which result in that a configuration of certain tenant id overrides the configuration at system level. 
+When the tenant level configuration is missing, fallback to the configuration at system level.
+- `TENANT_USER` level configuration overrides `TENANT` level, which result in a configuration of certain tenant id and username overrides the configuration at tenant level. 
+When the tenant user level configuration is missing, fallback to the configuration at tenant level.
+
+## Config Service
```
Comment (AngersZhuuuu) [inline docs/developers/configuration.md:41]: Menu like this?
```
# Configuration
## Static Configuration
## Dynamic Configuration
### Configuration Level
### ConfigService
#### FS Config Service
#### DB Config Service

```

File: docs/developers/configuration.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+Config service provides the configuration management service with local cache for the static and dynamic configuration. 
+Meanwhile, `ConfigService` is a pluggable service interface of which implementation is based on various store backend. 
+The store backend of `ConfigService` is configured with `celeborn.dynamicConfig.store.backend`, which supports filesystem 
+and database store backend at present. If the store backend is not provided, it means that config service is disabled. 
+
+### FileSystem Config Service
+Filesystem config service supports the configuration specified in the dynamic configuration file. The dynamic configuration 
+file is configured with `celeborn.quota.configuration.path`, of which default path is `$CELEBORN_HOME/conf/dynamicConfig.yaml`.
```
Comment (RexXiong) [inline docs/developers/configuration.md:49]: The configuration path key has been changed

File: docs/developers/configuration.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    gmt_modify   timestamp NOT NULL,
+    PRIMARY KEY (id),
+    UNIQUE KEY `index_unique_tenant_config_key` (`cluster_id`, `tenant_id`, `user`, `config_key`)
+);
+```
+
+After the creation of configuration tables, the dynamic configuration of different config levels is specified via inserting
+a configuration record in corresponding config level table.
+
+```sql
```
Comment (RexXiong) [inline docs/developers/configuration.md:142]: Don't use the dummy data in ut test,  we can add some real cases for example?

File: docs/developers/configuration.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+Dynamic configuration is a type of configuration that can be changed dynamically at runtime as needed, which supports various 
+config levels including:
+
+- `SYSTEM`: The system configurations.
+- `TENANT`: The dynamic configurations of tenant id.
+- `TENANT_USER`: The dynamic configurations of tenant id and username.
+
+When applying dynamic configuration, the priority order of config levels is as follows: 
+
+- `SYSTEM` level configuration overrides the static configuration, which in turn overrides the configuration of `CelebornConf`.
```
Comment (RexXiong) [inline docs/developers/configuration.md:34]: We should tell Celeborn users which configuration key is dynamic and where to find them. Currently those configuration keys with dynamic property in Celeborn Conf support dynamic configuration and can change at runtime.

File: docs/developers/configuration.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+  level: TENANT
+  config:
+    [config_key]: [config_val]
+    ...
+  users:
+    # TENANT_USER level configuration
+      - name: [name]
+      config:
+        [config_key]: [config_val]
+        ...
```
Comment (AngersZhuuuu) [inline docs/developers/configuration.md:75]:
```
# TENANT level configuration
- tenantId: [tenant_id]
  level: TENANT
  config:
    [config_key]: [config_val]
    ...
  users:
      # TENANT_USER level configuration
      - name: [name]
      config:
        [config_key]: [config_val]
        ...
```

File: docs/developers/configuration.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+);
+
+# TENANT/TENANT_USER level configuration
+CREATE TABLE IF NOT EXISTS celeborn_cluster_tenant_config
+(
+    id           int          NOT NULL AUTO_INCREMENT,
+    cluster_id   int          NOT NULL,
+    tenant_id    varchar(255) NOT NULL,
+    level        varchar(255) NOT NULL COMMENT 'config level, valid level is TENANT,USER',
+    user         varchar(255) DEFAULT NULL COMMENT 'tenant sub user',
```
Comment (AngersZhuuuu) [inline docs/developers/configuration.md:139]: have change to name?

File: docs/developers/configuration.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+- `TENANT` level configuration overrides `SYSTEM` level, which result in that a configuration of certain tenant id overrides the configuration at system level. 
+When the tenant level configuration is missing, fallback to the dynamic configuration at system level.
+- `TENANT_USER` level configuration overrides `TENANT` level, which result in a configuration of certain tenant id and username overrides the configuration at tenant level. 
+When the tenant user level configuration is missing, fallback to the dynamic configuration at tenant level.
+
+## Config Service
+Config service provides the configuration management service with local cache for the static and dynamic configuration. 
+Meanwhile, `ConfigService` is a pluggable service interface of which implementation is based on various store backend. 
+The store backend of `ConfigService` is configured via `celeborn.dynamicConfig.store.backend`, which supports filesystem(`FS`) 
+and database(`DB`) store backend. If the store backend is not provided, it means that config service is disabled.
```
Comment (AngersZhuuuu) [inline docs/developers/configuration.md:51]: `which supports filesystem(`FS`) and database(`DB`) store backend at present.`

File: docs/developers/configuration.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    [config_key]: [config_val]
+    ...
+  users:
+    # TENANT_USER level configuration
+      - name: [name]
+      config:
+        [config_key]: [config_val]
+        ...
+```
+For example, a Celeborn worker has 10 storage directories or disks and the buffer size is set to 256 KiB. The tenantId1
```
Comment (AngersZhuuuu) [inline docs/developers/configuration.md:77]:
```
`tenantId1`
```

File: docs/developers/configuration.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+only uses half of the storage and sets the buffer size to 128 KiB. Meanwhile, the user1 needs to change the buffer size
```
Comment (AngersZhuuuu) [inline docs/developers/configuration.md:78]:
```
`user1`
```

File: docs/developers/configuration.md [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+      config:
+       celeborn.worker.flusher.buffer.size: 96K # sets buffer size of tenantId1 and user1 to 128 KiB
+```
+
+### Database Config Service
+Database config service refreshes the dynamic configuration stored in the database via JDBC way. The configurations of 
+database store backend are specified with `celeborn.dynamicConfig.store.db.*` series. The database of config store backend 
+needs to create the tables of the dynamic configuration for different config levels.  The template of MySQL configuration 
+tables is as follows:
+
```
[the remark this answers is not in the document]
Comment (RexXiong) [reply docs/developers/configuration.md:102]: > Better mention `service/src/main/resources/sql/mysql/celeborn-0.5.0-mysql.sql`

This script is located at `db-scripts` directory in release package.

```# Copy db scripts
mkdir "$DIST_DIR/db-scripts"
cp -r "$PROJECT_DIR/service/src/main/resources/sql/" "$DIST_DIR/db-scripts"
```

File: docs/developers/configuration.md [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+    # TENANT_USER level configuration
+    - name: user1
+      config:
+        celeborn.worker.flusher.buffer.size: 96K # sets buffer size of tenantId1 and user1 to 128 KiB
+```
+
+### Database Config Service
+The database config service updates dynamic configurations stored in the database using the JDBC approach. Configuration settings for the database storage backend 
+are defined by the `celeborn.dynamicConfig.store.db.*` series of configuration keys. To use the database as a config store backend, it is necessary to create tables for 
+dynamic configurations at the various configuration levels. The template for MySQL configuration tables is structured as follows:
```
Comment (pan3793) [inline docs/developers/configuration.md:101]: DDL may change over time, should we expose these internal implementation details to user documentation? I suppose we are going to provide an editor on the Web UI eventually?

**Review by AngersZhuuuu — APPROVED** [review]
LGTM

**Comment by SteNicholas** [comment]
LGTM. Mreged into main(v0.5.0).

**Pull request closed without merging**